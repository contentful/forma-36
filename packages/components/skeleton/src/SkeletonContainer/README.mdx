---
title: 'SkeletonContainer'
type: 'component'
status: 'stable'
slug: /components/skeleton/skeleton-container/
github: 'https://github.com/contentful/forma-36/tree/master/packages/components/skeleton'
storybook: 'https://f36-storybook.contentful.com/?path=/story/components-skeleton-container--default'
typescript: ./SkeletonContainer.tsx
---

import { Props } from '@contentful/f36-docs-utils';

## How to use SkeletonContainer

The `SkeletonContainer` component is a wrapper component of the all other skeleton components: `SkeletonBodyText`, `SkeletonDisplayText`, `SkeletonImage`, `SkeletonRow`

Use can use properties of `SkeletonContainer` to control color, background color, animation, opacity of the skeleton elements that are used inside.

### Import

```js static=true
import { SkeletonContainer } from '@contentful/f36-components';
// or
import { SkeletonContainer } from '@contentful/f36-skeleton';
```

## Code examples

```jsx
<SkeletonContainer>
  <SkeletonBodyText numberOfLines={4} />
</SkeletonContainer>
```

An example with a different color and an increased speed.

```jsx
<SkeletonContainer backgroundColor="red" speed={1}>
  <SkeletonBodyText numberOfLines={2} />
</SkeletonContainer>
```

## Props

<Props of="SkeletonContainer" />
