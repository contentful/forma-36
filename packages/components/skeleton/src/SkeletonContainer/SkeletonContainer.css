@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';

.SkeletonContainer {
  display: block;
}
