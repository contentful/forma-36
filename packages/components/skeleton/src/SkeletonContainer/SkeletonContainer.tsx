import React from 'react';
import { Box, useId } from '@contentful/f36-core';
import type { CommonProps } from '@contentful/f36-core';

export interface SkeletonContainerProps extends CommonProps {
  /**
   * Background color of the skeleton
   */
  backgroundColor?: string;
  /**
   * Background opacity of the skeleton
   */
  backgroundOpacity?: number;
  /**
   * Whether skeleton has animation or not
   */
  isAnimated?: boolean;
  /**
   * Speed of the animation
   */
  speed?: number | string;
  /**
   * Color of the foreground skeleton items
   */
  foregroundColor?: string;
  /**
   * Opacity of the foreground skeleton items
   */
  foregroundOpacity?: number;
  /**
   * Width of the SVG element
   */
  svgWidth?: string | number;
  /**
   * Height of the SVG element
   */
  svgHeight?: string | number;
  /**
   * Label attribute
   */
  ariaLabel?: string;
  width?: number | string;
  height?: number | string;
  clipId?: string;
  gradientId?: string;
  animateId?: string;
  preserveAspectRatio?: string;
  children: React.ReactNode;
}

export const SkeletonContainer = ({
  children,
  testId = 'cf-ui-skeleton-form',
  ariaLabel = 'Loading component...',
  width = '100%',
  height = '100%',
  preserveAspectRatio,
  backgroundColor = '#e5ebed',
  backgroundOpacity = 1,
  isAnimated = true,
  speed = 2,
  foregroundColor = '#f7f9fa',
  foregroundOpacity = 1,
  svgWidth = '100%',
  svgHeight = '100%',
  clipId,
  gradientId,
  animateId,
  ...otherProps
}: SkeletonContainerProps) => {
  const uniqueClipId = useId(clipId, 'cf-ui-skeleton-clip');
  const uniqueGradientId = useId(gradientId, 'cf-ui-skeleton-clip-gradient');
  const randomAnimateId = useId(undefined, 'animation');
  const uniqueAnimateId = animateId || randomAnimateId;

  return (
    <Box
      as="svg"
      display="block"
      role="img"
      aria-label={ariaLabel}
      preserveAspectRatio={preserveAspectRatio}
      width={svgWidth}
      height={svgHeight}
      testId={testId}
      {...otherProps}
    >
      {ariaLabel ? <title>{ariaLabel}</title> : null}
      <rect
        x="0"
        y="0"
        width={width}
        height={height}
        clipPath={`url(#${uniqueClipId})`}
        style={{ fill: `url(#${uniqueGradientId})` }}
      />

      <defs>
        <clipPath id={uniqueClipId}>{children}</clipPath>

        <linearGradient id={uniqueGradientId}>
          <stop
            offset="0%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {isAnimated && (
              <animate
                id={uniqueAnimateId}
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="50%"
            stopColor={foregroundColor}
            stopOpacity={foregroundOpacity}
          >
            {isAnimated && (
              <animate
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                begin={`${uniqueAnimateId}.begin+0.25s`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="100%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {isAnimated && (
              <animate
                attributeName="stop-color"
                begin={`${uniqueAnimateId}.begin+0.5s`}
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
        </linearGradient>
      </defs>
    </Box>
  );
};

SkeletonContainer.displayName = 'SkeletonContainer';
