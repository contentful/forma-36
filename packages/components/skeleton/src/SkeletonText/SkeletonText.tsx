import React, { useCallback } from 'react';

type stringOrNumber = string | number;

export interface SkeletonTextProps {
  /**
   * A number of skeleton likes
   */
  numberOfLines?: number;
  /**
   * A distance between top of the container and the first line (in pixels)
   */
  offsetTop?: stringOrNumber;
  /**
   * A distance between left of the container and the beginning of lines (in pixels)
   */
  offsetLeft?: stringOrNumber;
  /**
   * A height of a one line (in pixels)
   */
  lineHeight?: stringOrNumber;
  /**
   * Spacing between lines (in pixels)
   */
  marginBottom?: stringOrNumber;
  /**
   * A width of a line
   */
  width?: stringOrNumber;
}

export const SkeletonText = ({
  numberOfLines = 1,
  offsetLeft = 0,
  offsetTop = 0,
  lineHeight = 21,
  marginBottom = 20,
  width,
}: SkeletonTextProps) => {
  const getLineWidth = useCallback(
    (lastLine: boolean) => {
      if (width) {
        return width;
      }

      return lastLine ? '80%' : '100%';
    },
    [width],
  );

  return (
    <React.Fragment>
      {Array.from(Array(numberOfLines)).map((_, index) => (
        <rect
          key={`skeleton-display-text-${index}`} // eslint-disable-line
          x={offsetLeft}
          y={
            index * (+lineHeight! + +marginBottom!) + +offsetTop! // eslint-disable-line @typescript-eslint/no-non-null-assertion
          }
          rx="0"
          ry="0"
          width={getLineWidth(
            numberOfLines! > 1 && numberOfLines! - index === 1, // eslint-disable-line @typescript-eslint/no-non-null-assertion
          )}
          height={lineHeight}
        />
      ))}
    </React.Fragment>
  );
};

SkeletonText.displayName = 'SkeletonText';
