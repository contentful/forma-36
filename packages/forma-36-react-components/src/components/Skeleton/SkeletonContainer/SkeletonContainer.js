import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export class SkeletonContainer extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    ariaLabel: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    preserveAspectRatio: PropTypes.string,
    clipId: PropTypes.string,
    gradientId: PropTypes.string,
    backgroundColor: PropTypes.string,
    backgroundOpacity: PropTypes.number,
    animate: PropTypes.bool,
    speed: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    foregroundColor: PropTypes.string,
    foregroundOpacity: PropTypes.number,
    svgWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    svgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  };

  static defaultProps = {
    extraClassNames: undefined,
    testId: 'cf-ui-skeleton-form',
    ariaLabel: 'Loading component...',
    width: '100%',
    height: '100%',
    preserveAspectRatio: 'xMidYMid meet',
    clipId: 'cf-ui-skeleton-clip-id',
    gradientId: 'cf-ui-skeleton-clip-gradient',
    backgroundColor: '#e5ebed',
    backgroundOpacity: 1,
    animate: true,
    speed: 2,
    foregroundColor: '#f7f9fa',
    foregroundOpacity: 1,
    svgWidth: '100%',
    svgHeight: '100%',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      ariaLabel,
      width,
      height,
      preserveAspectRatio,
      clipId,
      gradientId,
      backgroundColor,
      backgroundOpacity,
      animate,
      speed,
      foregroundColor,
      foregroundOpacity,
      svgWidth,
      svgHeight,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.SkeletonContainer, extraClassNames);

    return (
      <svg
        role="img"
        className={classNames}
        aria-label={ariaLabel}
        preserveAspectRatio={preserveAspectRatio}
        width={svgWidth}
        height={svgHeight}
        {...otherProps}
      >
        {ariaLabel ? <title>{ariaLabel}</title> : null}
        <rect
          x="0"
          y="0"
          width={width}
          height={height}
          clipPath={`url(#${clipId})`}
          style={{ fill: `url(#${gradientId})` }}
        />

        <defs>
          <clipPath id={clipId}>{children}</clipPath>

          <linearGradient id={gradientId}>
            <stop
              offset="0%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  id="animation1"
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="50%"
              stopColor={foregroundColor}
              stopOpacity={foregroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  begin="animation1.begin+0.25s"
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="100%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  begin="animation1.begin+0.5s"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
          </linearGradient>
        </defs>
      </svg>
    );
  }
}

export default SkeletonContainer;
