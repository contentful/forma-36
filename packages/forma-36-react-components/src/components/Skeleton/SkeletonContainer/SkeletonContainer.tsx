import React from 'react';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export interface SkeletonContainerProps {
  width?: number | string;
  height?: number | string;
  preserveAspectRatio?: string;
  clipId?: string;
  gradientId?: string;
  animateId?: string;
  backgroundColor?: string;
  backgroundOpacity?: number;
  animate?: boolean;
  speed?: number | string;
  foregroundColor?: string;
  foregroundOpacity?: number;
  svgWidth?: string | number;
  svgHeight?: string | number;

  ariaLabel?: string;
  className?: string;
  testId?: string;
  children: React.ReactNode;
}

let idCounter = 0;

export function SkeletonContainer({
  className,
  children,
  testId = 'cf-ui-skeleton-form',
  ariaLabel = 'Loading component...',
  width = '100%',
  height = '100%',
  preserveAspectRatio = 'xMidYMid meet',
  clipId = `cf-ui-skeleton-clip-${idCounter++}`,
  gradientId = `cf-ui-skeleton-clip-gradient-${idCounter++}`,
  animateId = `animation-${idCounter++}`,
  backgroundColor = '#e5ebed',
  backgroundOpacity = 1,
  animate = true,
  speed = 2,
  foregroundColor = '#f7f9fa',
  foregroundOpacity = 1,
  svgWidth = '100%',
  svgHeight = '100%',
  ...otherProps
}: SkeletonContainerProps): React.ReactElement {
  const classNames = cn(styles['SkeletonContainer'], className);

  return (
    <svg
      role="img"
      className={classNames}
      aria-label={ariaLabel}
      preserveAspectRatio={preserveAspectRatio}
      width={svgWidth}
      height={svgHeight}
      data-test-id={testId}
      {...otherProps}
    >
      {ariaLabel ? <title>{ariaLabel}</title> : null}
      <rect
        x="0"
        y="0"
        width={width}
        height={height}
        clipPath={`url(#${clipId})`}
        style={{ fill: `url(#${gradientId})` }}
      />

      <defs>
        <clipPath id={clipId}>{children}</clipPath>

        <linearGradient id={gradientId}>
          <stop
            offset="0%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                id={animateId}
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="50%"
            stopColor={foregroundColor}
            stopOpacity={foregroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                begin={`${animateId}.begin+0.25s`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="100%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                begin={`${animateId}.begin+0.5s`}
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
        </linearGradient>
      </defs>
    </svg>
  );
}
