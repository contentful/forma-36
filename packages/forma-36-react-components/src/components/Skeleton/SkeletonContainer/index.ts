export * from './SkeletonContainer';
export { default } from './SkeletonContainer';
