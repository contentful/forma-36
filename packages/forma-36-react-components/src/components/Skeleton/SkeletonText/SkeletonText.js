import React from 'react';
import PropTypes from 'prop-types';

export const SkeletonTextPropTypes = {
  numberOfLines: PropTypes.number,
  width: PropTypes.number,
  offsetTop: PropTypes.number,
  offsetLeft: PropTypes.number,
  lineHeight: PropTypes.number,
  marginBottom: PropTypes.number,
};

export class SkeletonText extends React.Component {
  static propTypes = SkeletonTextPropTypes;

  static defaultProps = {
    numberOfLines: 1,
    width: undefined,
    offsetTop: 0,
    offsetLeft: 0,
    lineHeight: 21,
    marginBottom: 20,
  };

  getLineWidth = lastLine => {
    if (this.props.width) {
      return this.props.width;
    }
    return lastLine ? '80%' : '100%';
  };

  render() {
    const {
      numberOfLines,
      offsetLeft,
      offsetTop,
      lineHeight,
      marginBottom,
    } = this.props;

    return (
      <React.Fragment>
        {Array.from(Array(numberOfLines)).map((_, index) => (
          <rect
            key={`skeleton-display-text-${index}`} // eslint-disable-line
            x={offsetLeft}
            y={index * (lineHeight + marginBottom) + offsetTop}
            rx="0"
            ry="0"
            width={this.getLineWidth(
              numberOfLines > 1 && numberOfLines - index === 1,
            )}
            height={lineHeight}
          />
        ))}
      </React.Fragment>
    );
  }
}

export default SkeletonText;
