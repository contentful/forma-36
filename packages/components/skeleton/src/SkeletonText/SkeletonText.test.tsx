import React from 'react';
import { render } from '@testing-library/react';
import { axe } from '@/scripts/test/axeHelper';

import { SkeletonContainer } from '../SkeletonContainer/SkeletonContainer';
import { SkeletonText } from './SkeletonText';

it('renders the component', () => {
  const { container } = render(
    <SkeletonContainer>
      <SkeletonText />
    </SkeletonContainer>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('renders 3 lines of skeleton text', () => {
  const { container } = render(
    <SkeletonContainer>
      <SkeletonText numberOfLines={3} />
    </SkeletonContainer>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(
    <SkeletonContainer>
      <SkeletonText />
    </SkeletonContainer>,
  );
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});
