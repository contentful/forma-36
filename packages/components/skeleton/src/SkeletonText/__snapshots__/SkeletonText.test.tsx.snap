// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 3 lines of skeleton text 1`] = `
<svg
  aria-label="Loading component..."
  class="SkeletonContainer"
  data-test-id="cf-ui-skeleton-form"
  height="100%"
  preserveAspectRatio="xMidYMid meet"
  role="img"
  width="100%"
>
  <title>
    Loading component...
  </title>
  <rect
    clip-path="url(#cf-ui-skeleton-clip-3)"
    height="100%"
    style="fill: url(#cf-ui-skeleton-clip-gradient-4);"
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clippath
      id="cf-ui-skeleton-clip-3"
    >
      <rect
        height="21"
        rx="0"
        ry="0"
        width="100%"
        x="0"
        y="0"
      />
      <rect
        height="21"
        rx="0"
        ry="0"
        width="100%"
        x="0"
        y="41"
      />
      <rect
        height="21"
        rx="0"
        ry="0"
        width="80%"
        x="0"
        y="82"
      />
    </clippath>
    <lineargradient
      id="cf-ui-skeleton-clip-gradient-4"
    >
      <stop
        offset="0%"
        stop-color="#e5ebed"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          dur="2s"
          id="animation-5"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="50%"
        stop-color="#f7f9fa"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          begin="animation-5.begin+0.25s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="100%"
        stop-color="#e5ebed"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          begin="animation-5.begin+0.5s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
    </lineargradient>
  </defs>
</svg>
`;

exports[`renders the component 1`] = `
<svg
  aria-label="Loading component..."
  class="SkeletonContainer"
  data-test-id="cf-ui-skeleton-form"
  height="100%"
  preserveAspectRatio="xMidYMid meet"
  role="img"
  width="100%"
>
  <title>
    Loading component...
  </title>
  <rect
    clip-path="url(#cf-ui-skeleton-clip-0)"
    height="100%"
    style="fill: url(#cf-ui-skeleton-clip-gradient-1);"
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clippath
      id="cf-ui-skeleton-clip-0"
    >
      <rect
        height="21"
        rx="0"
        ry="0"
        width="100%"
        x="0"
        y="0"
      />
    </clippath>
    <lineargradient
      id="cf-ui-skeleton-clip-gradient-1"
    >
      <stop
        offset="0%"
        stop-color="#e5ebed"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          dur="2s"
          id="animation-2"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="50%"
        stop-color="#f7f9fa"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          begin="animation-2.begin+0.25s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="100%"
        stop-color="#e5ebed"
        stop-opacity="1"
      >
        <animate
          attributeName="stop-color"
          begin="animation-2.begin+0.5s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
    </lineargradient>
  </defs>
</svg>
`;
